src/OpenApi/src/Transformers/Implementations/OpenApiSchemaReferenceTransformer.cs

```diff
@@ -0,0 +1,148 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Linq;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.OpenApi.Models;
+
+namespace Microsoft.AspNetCore.OpenApi;
+
+/// <summary>
+/// Document transformer to support mapping duplicate JSON schema instances
+/// into JSON schema references across the document.
+/// </summary>
+internal sealed class OpenApiSchemaReferenceTransformer : IOpenApiDocumentTransformer
+{
+    public Task TransformAsync(OpenApiDocument document, OpenApiDocumentTransformerContext context, CancellationToken cancellationToken)
+    {
+        var schemaStore = context.ApplicationServices.GetRequiredKeyedService<OpenApiSchemaStore>(context.DocumentName);
+        var schemasByReference = schemaStore.SchemasByReference;
+
+        document.Components ??= new OpenApiComponents();
+        document.Components.Schemas ??= new Dictionary<string, OpenApiSchema>();
+
+        foreach (var (schema, referenceId) in schemasByReference.OrderBy(kvp => kvp.Value))
```

Comment:
Nit: We could save some ordering.

```suggestion
        foreach (var (schema, referenceId) in schemasByReference.Where(kvp.Value is not null).OrderBy(kvp => kvp.Value))
```